### PR TITLE
fixes #1041 re #1041: TLDEventDataFilter should handle comparing two identifiers

### LINKED ISSUE #1041: TLDEventDataFilter should handle JEXL function for comparing two identifiers
Currently, TLDEventDataFilter/EventDataQueryExpressionFilter cannot handle JEXL function containing a comparison between two identifiers. It runs JexlASTHelper.getIdentifierOpLiteral on the node which cannot be resolved and throws NPE. We need to update the filter to handle this case properly.

FIELD_A:somevalue AND #JEXL("$FIELD_B == $FIELD_C")

### PR DESCRIPTION
**Pull request by d-hwang**
I tried to make local/safer changes for the code and the test. Please let me know if anyone has any concerns. Thanks!

### CODE REVIEW

**Review by FineAndDandy — APPROVED** [review]

**Review by ivakegg — DISMISSED** [review]

**Comment by d-hwang** [comment]
This code is ready for another review. Thanks!!

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/FieldToFieldComparisonVisitor.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
+package datawave.query.jexl.visitors;
+
+import datawave.query.jexl.JexlASTHelper;
+import org.apache.commons.jexl2.parser.*;
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/FieldToFieldComparisonVisitor.java:4]: Single imports please

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/FieldToFieldComparisonVisitor.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+public class FieldToFieldComparisonVisitor extends RebuildingVisitor {
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/FieldToFieldComparisonVisitor.java:6]: Short description for what this visitor does, e.g. "will wrap field to field comparison nodes in evaluation only markers", and perhaps an example "FIELD_A == FIELD_B" -> "(\_eval\_) & (FIELD_A.... etc

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/FieldToFieldComparisonVisitor.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+     *
+     * @param node
+     * @return
+     */
+    protected JexlNode evaluationOnlyForFieldToFieldComparison(JexlNode node) {
+        int identifierNodes = 0;
+        
+        // check both sides of nodes and count the nodes with identifier(s)
+        for (int i = 0; i < node.jjtGetNumChildren(); i++) {
+            if (JexlASTHelper.getIdentifiers(node.jjtGetChild(i)).size() > 0) {
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/FieldToFieldComparisonVisitor.java:29]: Is it worth checking to see if this node has two children exactly? Based on the entry points below I don't think we would run into the case where num_children > 2.

Can you add a test for "FIELD_A == FIELD_B == FIELD_C"? If only to document that we do not support this case.

File: warehouse/query-core/src/test/java/datawave/query/jexl/JexlASTHelperTest.java [hunk 4]
```diff
@@ -122,6 +125,18 @@ public void test4() throws Exception {
         }
     }
     
+    @Test
+    public void testGetLiteralValueSafely() throws Exception {
+        ASTJexlScript query = JexlASTHelper.parseJexlQuery("FOO == BAR");
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/jexl/JexlASTHelperTest.java:130]: This is a nitpick, I would like to see "FOO == FOO2", given how frequently we use "FOO == 'bar'"

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/EventDataQueryExpressionVisitorTest.java [hunk 5]
```diff
@@ -216,6 +216,23 @@ public void test() throws Exception {
         assertNull(filter.get("BAR"));
     }
     
+    @Test
+    public void testFieldToFieldComparison() throws Exception {
+        String originalQuery = "FOO == BAR";
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/jexl/visitors/EventDataQueryExpressionVisitorTest.java:221]: Same comment as above with regards to using FOO, FOO2, FOO3...etc

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/ExecutableDeterminationVisitorTest.java [hunk 6]
```diff
@@ -5,13 +5,7 @@
...
 import datawave.query.util.MetadataHelper;
 import org.apache.accumulo.core.client.TableNotFoundException;
-import org.apache.commons.jexl2.parser.ASTJexlScript;
-import org.apache.commons.jexl2.parser.ASTReference;
-import org.apache.commons.jexl2.parser.ASTReferenceExpression;
-import org.apache.commons.jexl2.parser.JexlNode;
-import org.apache.commons.jexl2.parser.JexlNodes;
-import org.apache.commons.jexl2.parser.ParseException;
-import org.apache.commons.jexl2.parser.ParserTreeConstants;
+import org.apache.commons.jexl2.parser.*;
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/jexl/visitors/ExecutableDeterminationVisitorTest.java:8]: Single imports please

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/ExecutableDeterminationVisitorTest.java [hunk 7]
```diff
@@ -560,4 +554,11 @@ public void testAllOrNoneEmptyPortion() throws TableNotFoundException {
         
         verifyAll();
     }
+    
+    @Test
+    public void testEvaluationOnlyReferenceNode() throws TableNotFoundException, ParseException {
+        // TODO update this test before committing
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/jexl/visitors/ExecutableDeterminationVisitorTest.java:560]: update this?

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/FieldToFieldComparisonVisitorTest.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
+package datawave.query.jexl.visitors;
+
+import datawave.query.jexl.JexlASTHelper;
+import org.apache.commons.jexl2.parser.ASTEvaluationOnly;
+import org.apache.commons.jexl2.parser.ASTJexlScript;
+import org.apache.commons.jexl2.parser.ParseException;
+import org.junit.Test;
+
+import static org.junit.Assert.*;
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/jexl/visitors/FieldToFieldComparisonVisitorTest.java:9]: single imports

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/FieldToFieldComparisonVisitorTest.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    @Test
+    public void testEqDoNothingFieldsToLiteral() throws ParseException {
+        ASTJexlScript query = FieldToFieldComparisonVisitor.forceEvaluationOnly(JexlASTHelper.parseJexlQuery("(FOO || BAR).min().hashCode() == 0"));
+        assertFalse(ASTEvaluationOnly.instanceOf(query));
+    }
+    
+    @Test
+    public void testEqDoNothing3() throws ParseException {
+        ASTJexlScript query = FieldToFieldComparisonVisitor.forceEvaluationOnly(JexlASTHelper.parseJexlQuery("(UUID =~ 'C.*?' || UUID =~ 'S.*?')"));
+        System.out.println(JexlStringBuildingVisitor.buildQuery(query));
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/jexl/visitors/FieldToFieldComparisonVisitorTest.java:33]: Remove print statement. 

These tests look good. A general comment when it comes to testing visitors that rewrite the query tree, in the past I've delegated to a test method like `test(query, expected)` , which then runs the visitor and makes any required asserts. The bonus there is you can visually see the input and expected output right, for example...
```
query = "FOO == FOO2"
expected = "((_eval_) && (FOO == FOO2))"
test(query, expected)
```
Not requesting this as a change, just offering it as an option in the future.

**Review by apmoriarty — CHANGES_REQUESTED** [review]

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/FieldToFieldComparisonVisitorTest.java (hunk 9 above)
Comment (d-hwang) [reply warehouse/query-core/src/test/java/datawave/query/jexl/visitors/FieldToFieldComparisonVisitorTest.java]: I removed the print statement. That label changed recently, so I didn't want to look for specific query. That's a good tip for writing unit tests in the future though. I will keep that in mind.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/FieldToFieldComparisonVisitor.java (hunk 3 above)
Comment (d-hwang) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/FieldToFieldComparisonVisitor.java:42]: I updated the unit test which just throws Parser exception. Thanks!

**Review by apmoriarty — APPROVED** [review]
LGTM

File: warehouse/query-core/src/main/java/datawave/query/jexl/JexlASTHelper.java [hunk 10]
```diff
@@ -493,6 +508,26 @@ public static JexlNode rereference(JexlNode node) {
         return node;
     }
     
+    public static List<IdentifierOpLiteral> getIdentifierOpLiteralForIdentifierToIdentifier(JexlNode node) {
```
Comment (FineAndDandy) [inline warehouse/query-core/src/main/java/datawave/query/jexl/JexlASTHelper.java:511]: javadoc please

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/EventDataQueryExpressionVisitor.java [hunk 11]
```diff
@@ -398,7 +398,20 @@ public Object visit(ASTFunctionNode node, Object data) {
...
+            generateValueFilter(iol, false);
+        } else {
+            List<JexlASTHelper.IdentifierOpLiteral> identifiers = JexlASTHelper.getIdentifierOpLiteralForIdentifierToIdentifier(node);
+            if (identifiers != null) {
+                for (JexlASTHelper.IdentifierOpLiteral identifier : identifiers) {
+                    generateValueFilter(identifier, true);
+                }
+            } else {
+                // unable to set filters. throw NPE to match generateValueFilter()
+                throw new NullPointerException("Null IdentifierOpLiteral");
```
Comment (FineAndDandy) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/EventDataQueryExpressionVisitor.java:412]: Does this result in a DatawaveQueryException?

**Review by FineAndDandy — CHANGES_REQUESTED** [review]
please add an end to end query test

**Review by FineAndDandy — DISMISSED** [review]

**Review by apmoriarty — APPROVED** [review]

**Review by FineAndDandy — APPROVED** [review]

**Pull request merged**